target-app/Home/Views/BottomSheetContainerViewController.swift

```diff
@@ -0,0 +1,196 @@
+                           delay: 0,
+                           usingSpringWithDamping: 0.8,
+                           initialSpringVelocity: 0.5,
+                           options: [.curveEaseOut],
+                           animations: {
+                self.view.layoutIfNeeded()
+            }, completion: { _ in
+                self.state = .initial
+            })
+        } else {
+            self.view.layoutIfNeeded()
+            self.state = .initial
+        }
+    }
+    
+    // MARK: - Pan Action
+    @objc func handlePan(_ sender: UIPanGestureRecognizer) {
+        let translation = sender.translation(in: bottomSheetViewController.view)
+        let velocity = sender.velocity(in: bottomSheetViewController.view)
+        
+        let yTranslationMagnitude = translation.y.magnitude
+        
+        switch sender.state {
+        case .began, .changed:
+            if self.state == .full {
+                guard translation.y > 0 else { return }
+                
+                topConstraint.constant = -(configuration.height - yTranslationMagnitude)
+                
+                self.view.layoutIfNeeded()
+            } else {
+                let newConstant = -(configuration.initialOffset + yTranslationMagnitude)
+                
+                guard translation.y < 0 else { return }
+                guard newConstant.magnitude < configuration.height else {
+                    self.showBottomSheet()
+                    return
+                }
+                
+                topConstraint.constant = newConstant
+                
+                self.view.layoutIfNeeded()
+            }
+        case .ended:
+            if self.state == .full {
+                
+                if velocity.y < 0 {
+                    // Bottom Sheet was full initially and the user tried to move it to the top
+                    self.showBottomSheet()
+                } else if yTranslationMagnitude >= configuration.height / 2 || velocity.y > 1000 {
+                    self.hideBottomSheet()
+                } else {
+                    self.showBottomSheet()
+                }
+            } else {
+                
+                if yTranslationMagnitude >= configuration.height / 2 || velocity.y < -1000 {
+                    
+                    self.showBottomSheet()
+                    
+                } else {
+                    self.hideBottomSheet()
+                }
+            }
+        case .failed:
+            if self.state == .full {
+                self.showBottomSheet()
+            } else {
+                self.hideBottomSheet()
+            }
+        default: break
+        }
+    }
+    
+    // MARK: - Configuration
+    public struct BottomSheetConfiguration {
+        let height: CGFloat
+        let initialOffset: CGFloat
+    }
+    
+    private let configuration: BottomSheetConfiguration
+    
+    // MARK: - State
+    public enum BottomSheetState {
```

Comment:
A good idea is to always have the internal /private types at the top of this declaration.

Reply:
done